### PR TITLE
[REV-1278] Create metadata table for SDN fallback

### PR DESCRIPTION
**Pull request by jhan217**
### What are we doing
We are adding a metadata table to keep track of the current state of the SDN fallback csv data in the database. 

## Created table:
```
mysql> describe payment_sdnfallbackmetadata;
+--------------------+--------------+------+-----+---------+----------------+
| Field              | Type         | Null | Key | Default | Extra          |
+--------------------+--------------+------+-----+---------+----------------+
| id                 | int(11)      | NO   | PRI | NULL    | auto_increment |
| created            | datetime(6)  | NO   |     | NULL    |                |
| modified           | datetime(6)  | NO   |     | NULL    |                |
| file_checksum      | varchar(255) | NO   |     | NULL    |                |
| download_timestamp | datetime(6)  | NO   |     | NULL    |                |
| import_timestamp   | datetime(6)  | YES  |     | NULL    |                |
| import_state       | varchar(255) | NO   | UNI | NULL    |                |
+--------------------+--------------+------+-----+---------+----------------+
7 rows in set (0.00 sec)
```

### Why are we doing this
In order to know the current state of the SDN fallback csv data in the database so we know whether to update the data or not.

Ticket: https://openedx.atlassian.net/browse/REV-1278

### Testing
-migration ran successfully on local machine
-automated unit tests

### CODE REVIEW

File: ecommerce/extensions/payment/models.py [hunk 1]
```diff
@@ -121,6 +122,25 @@ def _validate_percentage_value(self):
                 "Percentage greater than 100 not allowed."
             ))
 
+class SDNFallbackMetadata(models.Model):
+    """ Record metadata about the SDN fallback CSV file download """
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py]: According to https://github.com/edx/open-edx-proposals/pull/119/files the ID column should be a BigAutoField but I'm not particularly worried us hitting 2billion+ downloads anytime soon...

Could also use a UUID instead of an ID but apparently there's overhead with that/this should only every contain 3 rows max at a time so leaving the ID column with the default stuff, unless anyone has strong opinions.

File: ecommerce/extensions/payment/migrations/0027_auto_20200810_2158.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Generated by Django 2.2.15 on 2020-08-10 21:58
```
Comment (jhan217) [inline ecommerce/extensions/payment/migrations/0027_auto_20200810_2158.py]: Not sure why this file didn't already exist so I'm adding it here so that the next migration can pass...hope the timestamp on this one doesn't super matter...

File: ecommerce/extensions/payment/tests/test_models.py [hunk 3]
```diff
@@ -98,3 +99,26 @@ def test_validate_percentage_value_bad(self):
...
+        actual_metadata = SDNFallbackMetadata.objects.create(
+            file_checksum = self.file_checksum,
+            download_timestamp = self.download_timestamp
+        )
+
+        pytest.set_trace()
+
+        self.assertEqual(len(SDNFallbackMetadata.objects.all()), 1)
+
+       # self.assertEqual(actual_metadata.file_checksum = self.file_checksum)
```
Comment (jhan217) [inline ecommerce/extensions/payment/tests/test_models.py]: This stuff was failing because it thought `actual_metadata` was a `SyntaxError: keyword can't be an expression`

File: ecommerce/extensions/payment/tests/test_models.py [hunk 4]
```diff
@@ -98,3 +99,26 @@ def test_validate_percentage_value_bad(self):
             self.ecm.discount_value = value
             with self.assertRaises(ValidationError):
                 self.ecm.clean()
+
+
+class SDNFallbackMetadataTests(TestCase):
+    def setUp(self):
+        super(SDNFallbackMetadataTests, self).setUp()
```
Comment (jhan217) [inline ecommerce/extensions/payment/tests/test_models.py:109]: Modelled after the SDNCheckFailureTests but it's failing because I already have the table provisioned locally?

File: ecommerce/extensions/payment/migrations/0027_auto_20200811_1356.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Generated by Django 2.2.15 on 2020-08-11 13:56
```
Comment (jhan217) [inline ecommerce/extensions/payment/migrations/0027_auto_20200811_1356.py:1]: Keeping this migration in this file since they need to go out in order anyways, unless someone has a strong preference for this to be a separate PR. Enterprise confirmed that this migration should be safe to check in, per https://edx-internal.slack.com/archives/CD3L7U2SZ/p1597165347033400?thread_ts=1597163180.030500&cid=CD3L7U2SZ

File: ecommerce/extensions/payment/models.py [hunk 6]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
...
+
+    import_state = models.CharField(
+        max_length=255,
+        validators=[MinLengthValidator(1)],
+        unique=True,
+        choices=IMPORT_STATES,
+        default='New',
+    )
+
+    @classmethod
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py:154]: Open to suggestions on better ways of doing this. I made it a class method since it was operating on the entire dataset/rows that are not `self`. You can see examples of the invocation in the tests.

File: ecommerce/extensions/payment/models.py [hunk 7]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
...
+        ('New', 'New'),
+        ('Current', 'Current'),
+        ('Discard', 'Discard'),
+    ]
+
+    import_state = models.CharField(
+        max_length=255,
+        validators=[MinLengthValidator(1)],
+        unique=True,
+        choices=IMPORT_STATES,
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py:150]: Using `choices` seems to be the standard pattern for ENUMs types (as opposed to https://django-mysql.readthedocs.io/en/latest/model_fields/enum_field.html). I would have preferred using a DB constraint (e.g. the ENUM type in MySQL), since it appears that Django does the `choices` validation at the model level, but only if you explicitly ask it to validate the model (e.g. with `full_clean()`) or if it's an input coming from a UI form field. Otherwise Django will let you just save invalid models (see https://docs.djangoproject.com/en/3.1/ref/models/instances/#django.db.models.Model.clean_fields). Given that this is only going to be used for a specific use case & set in a script and likely won't have a front-end, I think not having the enum validation at the DB level should be ok.

File: ecommerce/extensions/payment/models.py [hunk 8]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
             ))
 
 
+class SDNFallbackMetadata(models.Model):
+    """ Record metadata about the SDN fallback CSV file download """
+    file_checksum = models.CharField(max_length=255, validators=[MinLengthValidator(1)])
+    download_timestamp = models.DateTimeField(auto_now_add=True)
+    import_timestamp = models.DateTimeField(null=True, blank=True)
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py:138]: Allowing this to be null since I'm assuming that the import timestamp will be added in after 1. the metadata row is created and 2. the csv data is loaded into the table

File: ecommerce/extensions/payment/models.py [hunk 9]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
             ))
 
 
+class SDNFallbackMetadata(models.Model):
+    """ Record metadata about the SDN fallback CSV file download """
+    file_checksum = models.CharField(max_length=255, validators=[MinLengthValidator(1)])
+    download_timestamp = models.DateTimeField(auto_now_add=True)
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py]: Not sure if we want the download_timestamp to be auto-added (which takes away our ability to modify it later) on row creation or if we want to manually set it (e.g. we download the file now and process it in 3 hours)

File: ecommerce/extensions/payment/tests/test_models.py [hunk 10]
```diff
@@ -98,3 +99,75 @@ def test_validate_percentage_value_bad(self):
...
+
+    def test_swap_non_existent_rows(self):
+        """The swap function shouldn't break / do anything if there are no existing rows"""
+
+        SDNFallbackMetadata.swap_states()
+        self.assertEqual(len(SDNFallbackMetadata.objects.all()), 0)
+
+    def test_swap_two_existing_rows(self):
+        """Swap the rows if they already exist"""
+        SDNFallbackMetadata.objects.create(
```
Comment (jhan217) [inline ecommerce/extensions/payment/tests/test_models.py:214]: Not sure where to place setup data as there are different setup cases here (some that require no rows in the db, some that require multiple rows in the db, etc.). IMO tests should only contain the code that needs to be tested, and not whatever setup code that's needed to create the proper state for the test.

File: ecommerce/extensions/payment/tests/test_models.py [hunk 11]
```diff
@@ -98,3 +99,75 @@ def test_validate_percentage_value_bad(self):
...
+
+        SDNFallbackMetadata.swap_states()
+
+        existing_a_metadata = SDNFallbackMetadata.objects.filter(file_checksum="A")[0]
+        self.assertEqual(existing_a_metadata.import_state, 'Current')
+        existing_b_metadata = SDNFallbackMetadata.objects.filter(file_checksum="B")[0]
+        self.assertEqual(existing_b_metadata.import_state, 'Discard')
+
+    def test_swap_all_rows_exist(self):
+        """The swap function should break & no changes should be made if there is an existing Discard row"""
```
Comment (jhan217) [inline ecommerce/extensions/payment/tests/test_models.py]: The A/Cs didn't include deleting any existing 'Discard' rows to avoid this case in the swap function, so I'm assuming that gets done elsewhere. It would make sense to put in this function, but if we did do that, then it'd be harder to test the atomic expectation / rollback of this function.

File: ecommerce/extensions/payment/models.py (hunk 9 above)
Comment (MatthewPiatetsky) [reply ecommerce/extensions/payment/models.py]: the way I understood it we're creating a row when we download the file so I think this makes sense

File: ecommerce/extensions/payment/models.py [hunk 12]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
             ))
 
 
+class SDNFallbackMetadata(models.Model):
+    """ Record metadata about the SDN fallback CSV file download """
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/payment/models.py]: I think this class could use a longer docstring talking about what this model is and how it works.
I would also link to the ticket (and in the ticket include a link to the ADR and any relevant google docs that we have)

File: ecommerce/extensions/payment/models.py [hunk 13]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
...
+        max_length=255,
+        validators=[MinLengthValidator(1)],
+        unique=True,
+        choices=IMPORT_STATES,
+        default='New',
+    )
+
+    @classmethod
+    @atomic
+    def swap_states(cls):
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/payment/models.py]: could we add a docstring to this method describing the business logic that this is implementing?

File: ecommerce/extensions/payment/models.py [hunk 14]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
...
+        existing_current_imports = SDNFallbackMetadata.objects.filter(import_state='Current')
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/payment/models.py]: why would there be more than one import?

File: ecommerce/extensions/payment/models.py [hunk 15]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
...
+        if existing_current_imports.exists():
+            existing_current = existing_current_imports[0]
+            existing_current.import_state = 'Discard'
+            existing_current.full_clean()
+            existing_current.save()
+
+        existing_new_imports = SDNFallbackMetadata.objects.filter(import_state='New')
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/payment/models.py]: why would there be more than one import?

File: ecommerce/extensions/payment/models.py [hunk 16]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
...
+        choices=IMPORT_STATES,
+        default='New',
+    )
+
+    @classmethod
+    @atomic
+    def swap_states(cls):
+        existing_current_imports = SDNFallbackMetadata.objects.filter(import_state='Current')
+        if existing_current_imports.exists():
+            existing_current = existing_current_imports[0]
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/payment/models.py]: minor comment but my understanding is that the more pythonic way to do this is to use .get and wrap that in a try except with the SDNFallbackMetadata .DoesNotExist exception

File: ecommerce/extensions/payment/models.py [hunk 17]
```diff
@@ -122,5 +124,43 @@ def _validate_percentage_value(self):
...
+            existing_current.import_state = 'Discard'
+            existing_current.full_clean()
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/payment/models.py]: what happens here if it fails validation?

File: ecommerce/extensions/payment/tests/test_models.py [hunk 18]
```diff
@@ -1,10 +1,11 @@
 # -*- coding: utf-8 -*-
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/payment/tests/test_models.py:1]: could you add a test for the validation error that is triggered by full_clean (i'm assuming you would check in the test that an exception is thrown and the database changes are rolled back)

**Review by MatthewPiatetsky — COMMENTED** [review]
I would squash the commits before we're done with the PR

File: ecommerce/extensions/payment/models.py (hunk 14 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: Is there a way to more efficiently get the one row? This was the first method I found that worked, so happy to change

File: ecommerce/extensions/payment/models.py (hunk 16 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: the `.exists()` call is to make sure that this method doesn't barf if there isn't a row with the import_state="Current"

File: ecommerce/extensions/payment/models.py (hunk 17 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: The caller can deal with the failed validation. The only real failed validation use case I see right now is if there's already an existing 'Discard' row, which would then fail the uniqueness constraint. This would also protect us against any additional validations we may want to add in the future (that, and I'm used to `save` functions validating the model before saving so you don't end up with invalid rows in the database).

File: ecommerce/extensions/payment/models.py (hunk 15 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: see comment above on changing this to a more efficient method...

File: ecommerce/extensions/payment/tests/test_models.py (hunk 18 above)
Comment (jhan217) [reply ecommerce/extensions/payment/tests/test_models.py:1]: Is there something more specific you'd like beyond https://github.com/edx/ecommerce/pull/3087/files/485d4d6eb2a4216fdaabb82261a5aa298177216c#diff-6aa12e1c48e74f5bcc0bd29fabdf3185R165 ?

**Comment by jhan217** [comment]
@MatthewPiatetsky Are you saying to squash all the commits before clicking the 'Squash and Merge' button? Generally I prefer to keep the commits on a PR separated so that functionality can get moved across prs or dropped if needed/make it easier for reviewers to look at one set of changes at a time if they'd prefer, and then use the `Squash and merge` option when merging so that only one commit makes it to master.

File: ecommerce/extensions/payment/models.py (hunk 7 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py:150]: Note that we aren't using Django's EnumField since 1) (according to Calen) we need to support multiple database backends (e.g. mysql and postgres) for OpenEdx instances and 2) it's not currently used in either ecommerce or edx-platform so whatever lift is needed to get django's EnumField/database-agnostic-enum-types implemented is out of scope for this ticket.

File: ecommerce/extensions/payment/models.py (hunk 9 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: That was my understanding as well, but the ticket implementation notes seemed to suggest something different, so I was a little confused

File: ecommerce/extensions/payment/models.py [hunk 19]
```diff
@@ -147,6 +151,11 @@ class SDNFallbackMetadata(models.Model):
     @classmethod
     @atomic
     def swap_states(cls):
+        """
+        Update the newly downloaded + imported SDN file data to be used and prepare the previously used
+        SDN file data to be removed. This is done in a transaction to gurantee that existing files are
+        shifted into the correct next states in sync.
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py]: Note: This function does not attempt (to the best of it's ability) to make sure that there is always a row in the 'Current' state. Happy to add that if needed, but it wasn't specified in the ticket.

File: ecommerce/extensions/payment/models.py (hunk 17 above)
Comment (MatthewPiatetsky) [reply ecommerce/extensions/payment/models.py]: I just meant I'm curious what would happen, does it throw an exception? This is a behavior I think we should test at least

File: ecommerce/extensions/payment/tests/test_models.py (hunk 18 above)
Comment (MatthewPiatetsky) [reply ecommerce/extensions/payment/tests/test_models.py:1]: i wasn't entirely following whether the test is already doing this, but could you check that if the validation error happens in the second half, the first half is reversed due to the atomic decorator

File: ecommerce/extensions/payment/models.py (hunk 16 above)
Comment (MatthewPiatetsky) [reply ecommerce/extensions/payment/models.py]: to clarify for this comment and the next one, the more pythonic way to do this would be with try/except like being done here https://github.com/edx/ecommerce/blob/a402e5c5c3baf6a5285f5a706770b846eb64e1c9/ecommerce/extensions/api/v2/views/vouchers.py#L78
had trouble finding an authoritative article, but here's some explanation of why
https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/

Comment (MatthewPiatetsky) [reply ecommerce/extensions/payment/models.py]: feel free to slack me if this comment doesn't make sense

File: ecommerce/extensions/payment/tests/test_models.py (hunk 18 above)
Comment (jhan217) [reply ecommerce/extensions/payment/tests/test_models.py:1]: That's what I was attempting to do with the assertions below...? Unless I'm doing it incorrectly?

Comment (jhan217) [reply ecommerce/extensions/payment/tests/test_models.py:1]: The intention of the linked test was to call `SDNFallbackMetadata.swap_states()`, make sure it throws the error, and then validate that anything that happened in the function call would be rolled back/check that the rows are in their original states.

Comment (MatthewPiatetsky) [reply ecommerce/extensions/payment/tests/test_models.py:1]: sounds good!

File: ecommerce/extensions/payment/models.py (hunk 17 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: I thought I was testing the behavior in https://github.com/edx/ecommerce/pull/3087/files/485d4d6eb2a4216fdaabb82261a5aa298177216c#diff-6aa12e1c48e74f5bcc0bd29fabdf3185R165, is there a different test you're suggesting?

File: ecommerce/extensions/payment/models.py (hunk 16 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: Happy to write it with the try/except syntax, though not sure it's would be particularly useful if the except block just passes along the error + test that that also triggers the rollback

File: ecommerce/extensions/payment/models.py (hunk 17 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: ~Oh I see, it seems like `with self.assertRaises(ValidationError)` doesn't actually check if the error was raised, you actually need to additionally do another assert `self.assertRaises(ExpectedException, afunction)` to check that it was actually thrown~

It looks like just the `with self.assertRaises` will fail if no exceptions are thrown and the additional context stuff is only needed if you need to check specific attributes on the ValidationError (e.g. the error text) per https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

File: ecommerce/extensions/payment/models.py (hunk 16 above)
Comment (MatthewPiatetsky) [reply ecommerce/extensions/payment/models.py]: writing it that way is considered more pythonic (more in the python coding standard), but people write stuff all sorts of ways in our code, so fixing it is not required

Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: ah ok, I'll leave it as is unless someone else has a better idea of what to put in the `except` block that's not just passing up the error...

**Review by MatthewPiatetsky — APPROVED** [review]

File: ecommerce/extensions/payment/tests/test_models.py (hunk 11 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/tests/test_models.py]: I'm with you! I agree that deleting the existing Discard row makes the most sense in this function (good to confirm with Emma if it's intended in the ACs). 

I'm just learning about atomicity, but if that deletion step is the first step in the swap_states try block, nothing would be committed until/unless all steps are exception-free, so the atomicity/rollback would work properly in a test too, right?

Comment (jhan217) [reply ecommerce/extensions/payment/tests/test_models.py]: From my understanding of general atomicity in programming (it may be different in Python) - yes, it's all or nothing. Either everything completes with no errors or nothing gets completed. If we discard the row as part of this function, then we wouldn't be able to trigger that validation error in the test (it's relying on the fact that we're trying to move a row to 'Discard' when there's already a 'Discard' row and you can't have more than one 'Discard' rows in the database), so I would need to find some other way of triggering an exception to make sure the rollback works in the test.

One reason to not delete the 'Discard' row in this function would be if there are some checks that need to be done before deleting the 'Discard' row and we'd like to encapsulate that in its own separate function (e.g. we need to make sure that there's a row in the 'Current' & a row in the 'New' status before deleting the 'Discard' row).

File: ecommerce/extensions/payment/models.py (hunk 9 above)
Comment (julianajlk) [reply ecommerce/extensions/payment/models.py]: Is there a time we would want to modify it after it was downloaded? If for example there was some error in the download, would it have to wait for the next download or could we manually trigger a re-download?

Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: Discussed with Emma, we'll be adding the normal created/last edited timestamps and leave download/import timestamps to be set by the script/job

Download timestamp - when the file was downloaded (we should always have this when a metadata row is created). I think this is the one that gets updated if the downloaded file is the same as the file we have already.
Import timestamp - when the contents of the file are done getting imported into the database.

File: ecommerce/extensions/payment/models.py [hunk 20]
```diff
@@ -124,14 +125,16 @@ def _validate_percentage_value(self):
...
 
 
-class SDNFallbackMetadata(models.Model):
+class SDNFallbackMetadata(TimeStampedModel):
     """
     Record metadata about the SDN fallback CSV file download, as detailed in docs/decisions/0007-sdn-fallback.rst
     and JIRA ticket REV-1278. This table is used to track the state of the SDN csv file data that are currently
     being used or about to be updated/deprecated. This table does not keep track of the SDN files over time.
     """
+    logger = logging.getLogger("SDNFallbackMetadataLogger")
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py]: I'm not actually sure if this is how you're supposed to do logging or how to test that it gets called in the tests, but it's not broken...

File: ecommerce/extensions/payment/models.py (hunk 19 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: Talked to Emma, we're going to attempt this.

File: ecommerce/extensions/payment/tests/test_models.py (hunk 11 above)
Comment (jhan217) [reply ecommerce/extensions/payment/tests/test_models.py]: Talked to Emma, added the discarding while swapping.

File: ecommerce/extensions/payment/tests/test_models.py (hunk 10 above)
Comment (jhan217) [reply ecommerce/extensions/payment/tests/test_models.py:214]: Punting this

File: ecommerce/extensions/payment/models.py (hunk 9 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: Added created/modified timestamps; download/import will be used for business logic now.

File: ecommerce/extensions/payment/models.py [hunk 21]
```diff
@@ -122,5 +125,82 @@ def _validate_percentage_value(self):
...
+        shifted into their next states in sync and tries to ensure that there is always a row
+        in the 'Current' state. Rollbacks of all row's import_state changes will happen if:
+        1) There are multiple rows & none of them are 'Current', or
+        2) There are any issues with the existing rows + updating them (e.g. a row with a
+        duplicate import_state is manually inserted into the table during the transaction)
+        """
+        SDNFallbackMetadata._swap_state('Discard')
+        SDNFallbackMetadata._swap_state('Current')
+        SDNFallbackMetadata._swap_state('New')
+        if len(SDNFallbackMetadata.objects.all()) > 1:
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py]: We're not failing on the case where there's only one row & it's the Discard row, but it seems like an edge case. If there are 2+ rows at this point, at least one of them should be in the 'Current' state.

File: ecommerce/extensions/payment/models.py [hunk 22]
```diff
@@ -122,5 +125,82 @@ def _validate_percentage_value(self):
...
+            try:
+                SDNFallbackMetadata.objects.get(import_state='Current')
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py:180]: Putting this in a try/catch block so we can log a more useful error.

File: ecommerce/extensions/payment/models.py [hunk 23]
```diff
@@ -122,5 +125,82 @@ def _validate_percentage_value(self):
...
+        """
+        try:
+            existing_metadata = SDNFallbackMetadata.objects.get(import_state=import_state)
+            if import_state == 'Discard':
+                existing_metadata.delete()
+            else:
+                if import_state == 'New':
+                    existing_metadata.import_state = 'Current'
+                elif import_state == 'Current':
+                    existing_metadata.import_state = 'Discard'
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py:202]: Happy to add an 'else' clause to deal with invalid import states, but if we're not manually manipulating rows it doesn't seem like it would be an issue.

File: ecommerce/extensions/payment/models.py (hunk 20 above)
Comment (emma-green) [reply ecommerce/extensions/payment/models.py]: Generally the convention is to put the logger definition outside of the class. `log = logging.getLogger(__name__)`

You can use LogCapture, and then check check for the presence of you log. Here's an example: https://github.com/edx/ecommerce/blob/master/ecommerce/core/tests/test_models.py#L113

File: ecommerce/extensions/payment/models.py (hunk 21 above)
Comment (emma-green) [reply ecommerce/extensions/payment/models.py]: there's no else that's associated with this if. What do you want to happen if there's only one row?

File: ecommerce/extensions/payment/models.py [hunk 24]
```diff
@@ -122,5 +125,82 @@ def _validate_percentage_value(self):
...
+        SDNFallbackMetadata._swap_state('Discard')
+        SDNFallbackMetadata._swap_state('Current')
+        SDNFallbackMetadata._swap_state('New')
+        if len(SDNFallbackMetadata.objects.all()) > 1:
+            try:
+                SDNFallbackMetadata.objects.get(import_state='Current')
+            except SDNFallbackMetadata.DoesNotExist:
+                SDNFallbackMetadata.logger.info(
+                    "Expected a row in the 'Current' import_state after swapping, but there are none",
+                )
```
Comment (emma-green) [inline ecommerce/extensions/payment/models.py:184]: should this be a warning? Not info?

File: ecommerce/extensions/payment/tests/test_models.py [hunk 25]
```diff
@@ -98,3 +99,139 @@ def test_validate_percentage_value_bad(self):
...
+            file_checksum="A",
+            import_state="New",
+            download_timestamp=self.download_timestamp,
+        )
+
+        SDNFallbackMetadata.swap_all_states()
+
+        actual_rows = SDNFallbackMetadata.objects.all()
+        self.assertEqual(len(actual_rows), 1)
+        self.assertEqual(actual_rows[0].import_state, 'Current')
```
Comment (emma-green) [inline ecommerce/extensions/payment/tests/test_models.py:144]: Is the current one always going to be first?

**Review by emma-green — COMMENTED** [review]
See my comments

File: ecommerce/extensions/payment/tests/test_models.py (hunk 25 above)
Comment (jhan217) [reply ecommerce/extensions/payment/tests/test_models.py:144]: Discussed with Emma, we'll keep this instead of using `get()` because we want to be explicit on what fails in this test (e.g. we'd like this test to fail its assertions, not because a method threw an exception). There is only one row that exists in this table, so ordering doesn't matter.

File: ecommerce/extensions/payment/models.py (hunk 21 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: Per discussion with Emma, will put in some code comments.

Comment (jhan217) [reply ecommerce/extensions/payment/models.py]: Actually, amending this to be `>= 1`. 
If there are 0 existing rows, swapping should do nothing, and therefore we can't have a 'Current' row.
If there is 1 existing row, we expect that row to start as 'New', then progress to 'Current' in this method. 
- If that row starts off as 'Current', we shouldn't move it to 'Discard' if there isn't a new file to replace it with.
- If that row starts off as 'Discard', we might as well delete it because we don't need it anymore and we shouldn't be in this state in the first place.

File: ecommerce/extensions/payment/models.py [hunk 26]
```diff
@@ -122,5 +127,87 @@ def _validate_percentage_value(self):
...
+        """
+        try:
+            existing_metadata = SDNFallbackMetadata.objects.get(import_state=import_state)
+            if import_state == 'Discard':
+                existing_metadata.delete()
+            else:
+                if import_state == 'New':
+                    existing_metadata.import_state = 'Current'
+                elif import_state == 'Current':
+                    existing_metadata.import_state = 'Discard'
```
Comment (jhan217) [inline ecommerce/extensions/payment/models.py:202]: If we move something to the 'Current' state, do we want to check that the import_timestamp is set? (happy to punt this to a different ticket)

**Review by emma-green — APPROVED** [review]

File: ecommerce/extensions/payment/models.py (hunk 26 above)
Comment (jhan217) [reply ecommerce/extensions/payment/models.py:202]: Emma: separate ticket
